Fix blob content type not parsed from queries

Rudiksz opened the pull request:
- Added public contructors to the Blob class
  Makes the Blob class a bit cleaner and allows creating Blob objects from outside the library (my service classes need to be able to create blobs)

- Added Document.fromMap constructor
   with this it's possible to create proper Document objects from query results. See examples in the gist below

- Added Database.initWithNameAndPath
   just for testing at the moment

All existing code using the Document and MutableDocument should be unaffected. See examples below for normal document read and queries including joins.

https://gist.github.com/Rudiksz/cd08fb0aecab3c0a2efa534bcc03419d

Rudiksz committed:
Fix blob content type not parsed from queries
Added public contructors to the Blob class
Added Document.fromMap constructor
Added Database.initWithNameAndPath

Rudiksz committed:
Added tests

bawelter commented:
There is an issue with changing the private methods to public, basically if you can't initialize the Blob with the way the accessors are already then we wont be able to on the platform code either.  I can't lookup an existing Blob with the JSON information being provided, I can only create a new one.  So if the only way you can create the blob is with the init function then that is how it should be created because this is what will happen in the platform code if there is no reference to the Blob Object from the original Document or Result Object on the platform code.  If you look at the Swift and Java libraries and think about what is available and how you would be able to recreate the Blob given the information you are passing from the JSON it might be easier to understand.

The only way to get by this would be to store the query to access the Blob on the platform code from what I am seeing.  You can see that is the way I get the data for Blobs in a document.  We can also store the Blob in a hash map where the key is the digest but we would need a way to clean up the hashmap when the blob is no longer needed using reference counting or something but we would need to make sure we this is cleaned up or we could run into memory issues.   The other hack is create a new Blob object every time a Blob is returned in a query which is basically the solution currently in this pull request.  Please let me know if you have any questions or thoughts.

It may be better to discuss solutions before implementation and doing a pull request especially since this more of an advanced solution and not just changing the key names in a json object.

Rudiksz commented:
I apreciate taking the time to review it, but I have to be honest, I don't understand what the issues are. 

I've been looking at the blobs for a week now and they work perfectly fine, if it wasn't for the content_type vs contentType mismatch between the dart code and the plugin platform code. Once I made sure the Blob.dart class reads to content_type properly (and therefore I can send it back to the platform code), blobs are saved, updated and deleted just as I would expect it. The blob metadata is returned both by 'getDocumentWithId' and doing a query using "SelectResult.all()", and I can parse it out from the map just fine. The content itself is lazy loaded whenever I call 'blob.content', usually in a futureBuilder widget.

I looked inside the database file itself and blobs are handled exactly as they should.

The issue I am having is that I am querying documents, instead of using 'getDocumentById()', and I need a way to get the Blob metadata out of the returned Result map, The Blob.dart class is not useful for that because it expects "contentType" in the returned map, while the Result has "content_type".
As a consequence when I'm updating a document and set back the blob, if the content_type is set to null, couchbase (or the plugin) deletes the blob (even if all other fields are present including the data). I don't know why, I don't know if it's because of the plugin's platform code or if it's couchbase doing it. Fixing the content type not being null, fixes the issue. Fixing the Blob.dart class to parse the map properly fixes the issue
 
My original solution was to fix the constructor from this:
```
  Blob._fromMap(Map<String, dynamic> map) {
    this._contentType = map["contentType"];
    this._digest = map["digest"];
    this._length = map["length"];
  }
```

to this
```
  Blob._fromMap(Map<String, dynamic> map) {
    this._contentType = map["contentType"] ?? map["content_type"];
    this._digest = map["digest"];
    this._length = map["length"];
  }
```
This fixes the problems I had with blobs disappearing and it's compatible with both existing code and the query results. 
Changing that one line of code is all that is needed for my app to work properly.
But as I said it's a hack. The platform code should send maps to the dart code as they are in the database. 


In this second pull request I changed the platform code to use "content_type" and be consistent with the couchbase core. The Blob.dart class reads the content_type both from Document class and Result class now, and the plugin saves, updates and deletes blobs exactly as it should.

Everything else I did is a cosmetic changes to make  the code a bit more user friendly.

It's literally just a key mismatch issue. I don't have any feelings about which way it should be fixed, but if neither soutions are good I will just have to write my own blob serialization/deserialization, so it doesn't get overwritten by future releases.

bawelter commented:
> I apreciate taking the time to review it, but I have to be honest, I don't understand what the issues are.
> 
> I've been looking at the blobs for a week now and they work perfectly fine, if it wasn't for the content_type vs contentType mismatch between the dart code and the plugin platform code. Once I made sure the Blob.dart class reads to content_type properly (and therefore I can send it back to the platform code), blobs are saved, updated and deleted just as I would expect it. The blob metadata is returned both by 'getDocumentWithId' and doing a query using "SelectResult.all()", and I can parse it out from the map just fine. The content itself is lazy loaded whenever I call 'blob.content', usually in a futureBuilder widget.
> 
> I looked inside the database file itself and blobs are handled exactly as they should.
> 
> The issue I am having is that I am querying documents, instead of using 'getDocumentById()', and I need a way to get the Blob metadata out of the returned Result map, The Blob.dart class is not useful for that because it expects "contentType" in the returned map, while the Result has "content_type".
> As a consequence when I'm updating a document and set back the blob, if the content_type is set to null, couchbase (or the plugin) deletes the blob (even if all other fields are present including the data). I don't know why, I don't know if it's because of the plugin's platform code or if it's couchbase doing it. Fixing the content type not being null, fixes the issue. Fixing the Blob.dart class to parse the map properly fixes the issue
> 
> My original solution was to fix the constructor from this:
> 
> ```
>   Blob._fromMap(Map<String, dynamic> map) {
>     this._contentType = map["contentType"];
>     this._digest = map["digest"];
>     this._length = map["length"];
>   }
> ```
> 
> to this
> 
> ```
>   Blob._fromMap(Map<String, dynamic> map) {
>     this._contentType = map["contentType"] ?? map["content_type"];
>     this._digest = map["digest"];
>     this._length = map["length"];
>   }
> ```
> 
> This fixes the problems I had with blobs disappearing and it's compatible with both existing code and the query results.
> Changing that one line of code is all that is needed for my app to work properly.
> But as I said it's a hack. The platform code should send maps to the dart code as they are in the database.
> 
> In this second pull request I changed the platform code to use "content_type" and be consistent with the couchbase core. The Blob.dart class reads the content_type both from Document class and Result class now, and the plugin saves, updates and deletes blobs exactly as it should.
> 
> Everything else I did is a cosmetic changes to make the code a bit more user friendly.
> 
> It's literally just a key mismatch issue. I don't have any feelings about which way it should be fixed, but if neither soutions are good I will just have to write my own blob serialization/deserialization, so it doesn't get overwritten by future releases.

So I guess the better way to think of this is that you are essentially not updating the document because the document(withID:) isn't being used.  You are creating a new document and overwriting it.  Without the old document the platform code doesn't know what the original Blob used to be when the data is null / the key content_type key is used.  This causes the Blob to not be entered into the new document entirely and is why you are having issues.  I am suggesting instead of making it look like the same Blob on the dart code, create a new Blob when the Blob is in a query.  This will not hide what is actually happening and will fix the issue as well.  This will also be easier to understand when debugging issues and looking through the code.

bawelter commented:
We should be able to make what you want to do work but I will need to change the platform code a bit.  It actually won't require changing much code but it will get rid of the lazy loading of Blobs in order to get around needing to keep a reference around for the Blobs.

Rudiksz commented:
I think we are talking about two different things. I don't know what lazy loading and reference counting has to do with changing the key of a json property.

I do my own parsing of the result map returned by the queries, now including the blobs too. Since I'm not relying on the Blob class anymore, all I need for the platform code to don't rewrite any of the json keys in the dictionary returned by the query.

Rudiksz commented:
> So I guess the better way to think of this is that you are essentially not updating the document because the document(withID:) isn't being used. You are creating a new document and overwriting it.

That's just not corrent. Couchbase lite does an update if you try to save a document with and ID that already exists in the database. Regardless if you pass revision or sequence number, couchbase lite treats every save as an update. If you try to save a document without revision and sequence using FAIL_ON_CONFLICT, the save will fail instead of deleting the previous document and recreating a new one. If you use LAST_WRITE_WINS, a new revision is created in the existing document.

> Without the old document the platform code doesn't know what the original Blob used to be when the data is null / the key content_type key is used.

Well, yes, the platform code has different logic to save a document, based on the sequence number, which makes the plugin work different from what the offical sdk does, and probably source of some bugs.

Deleting blobs when content_type is null is one of them. 
ConcurrencyControl variable of the database.save method in dart being broken is probably another.
If the platform code would simply forward the saveDocument call to the SDK instead of trying to do concurrency control, probably all these bugs would disappear.

bawelter commented:
So I created a new branch with the fixes on them that should meet your needs and not cause any breaking changes for other users. I did test what you told me on Couchbase creating a new revision in the existing document and that is true.  I remember there being an issue with doing that way for something but maybe I'm just going crazy lol.  I'm going to close this pull request and merge the new branch which I tested on both Android and iOS. 

There is also a new Fragment object that I added which will help with getting the Objects that are deeper in the document.  You can access these Fragments now with new operators on the Result, Document, and MutableDocument object such as myResult['userProfile']['gravatar'].getBlob() to get the user gravatar.

Please let me know if there are any issues.

bawelter commented:
If you have any issues with the updates I made to fix these issues please let me know so we can try to figure out a solution.

Rudiksz commented:
Looks allright, but the blob data being loaded is a deal breaker for me. Removing the lazy loading made the syntax cleaner, which is nice, but at the cost of being very inefficient.

I have many instances where I load a large list of documents (1-200) and each might have a couple of blobs. In a simple list/detail pattern I display only simple list tiles while the images would only displayed when the user taps a tile.

I don't like the idea of loading hundreds of blobs that may or may not get displayed. Besides the wasteful memory usage, parsing query  results now causes a roundtrip to the platform code for every blob in the result set. While it's fast, it's very wasteful. Live queries make it even worse, because they return the whole result set again instead of the only the changed document, which would cause to get the same hundreds of blobs again and again.

I'd rather have the blobs load their content on demand as it was before, potentially caching it on the dart side. Similar in concept to what the cached_network_image package does.

bawelter commented:
So the content is only loaded if you actually use the getBlob function otherwise the file is never loaded into memory.  The current implementation was done this way so that if you would like to improve the code without touching the platform code you can now do that.  I only load it like I am because I didn't have time to make a perfect solution and this solution should work in all scenarios.  If you would like to improve on the solution that would be great!

bawelter commented:
> Looks allright, but the blob data being loaded is a deal breaker for me. Removing the lazy loading made the syntax cleaner, which is nice, but at the cost of being very inefficient.
> 
> I have many instances where I load a large list of documents (1-200) and each might have a couple of blobs. In a simple list/detail pattern I display only simple list tiles while the images would only displayed when the user taps a tile.
> 
> I don't like the idea of loading hundreds of blobs that may or may not get displayed. Besides the wasteful memory usage, parsing query results now causes a roundtrip to the platform code for every blob in the result set. While it's fast, it's very wasteful. Live queries make it even worse, because they return the whole result set again instead of the only the changed document, which would cause to get the same hundreds of blobs again and again.
> 
> I'd rather have the blobs load their content on demand as it was before, potentially caching it on the dart side. Similar in concept to what the cached_network_image package does.

If you would like to change this back to lazy loading that shouldn't be an issue, just pull request it and keep the changes internal so the syntax doesn't change.  If you only change the dart code we shouldn't need to test on iOS.

bawelter commented:
Try the beta branch, I changed it to lazy load.

Rudiksz commented:
Thanks, it works as before.
Note though that with this the blob cache in the platformcode can not be emptied anymore, which will cause the memory usage to slowly creep up.  If some part of my app deletes the blob cache (which I would want to do once in a while), the blobs data can't be loaded.

Unless Couchbase adds a method to their Database class to get a blob by the digest, (which they probably could) the way blobs were handled before was the least wasteful way to load them. Yeah, the Blob class in dart having to keep the document id and database name within themselves and having to do extra document read on the platform side is somewhat annoying, but it was a tradeoff that made sense.

I'm writing an app that runs for many hours at end, so it's somewhat a concerning to cache things on the platform side where I can't easily dispose of objects.

I'm ok with the change for now, but it's definitely something that should be looked into.

bawelter commented:
I agree this does need looked into and I'm not really sure how the platform code handles the memory with the blobs. We can definitely look into better solutions. This is a starting point though and allows more flexibility to change the dart code to reach that solution.  I can always add another method to delete specific blobs in the cache so we can store them in a cache on the dart side but the best solution may be figuring out a way to access the blobs files directly so we don't need to transfer the file over the platform channel at all.

Rudiksz commented:
> but the best solution may be figuring out a way to access the blobs files directly so we don't need to transfer the file over the platform channel at all.

That's an interesting idea. If the Flutter app has access to the directory where the database is stored -which it should have, because by default it's stored in the app's data directory-, maybe we could read the files directly from the Attachments folder.

It would make it dependant on couchbase lite not changing how and where they store blobs, because that would cause blobs to break. I don't think they would ever do it between minor versions, and even if they ever do, the plugin could handle the differences.

Currently the files are stored in the Attachments folder using the digest for names.
It should be doable, I'll do some tests.

Edit: come to think of it, that's probably what SDK's getContent does too. Just reads the file from the disk.
We could read the blobs in the memory traditionally or get them as streams too, just like the native api does.
